Insert from web - allowing YouTube URLs

ranjanjharavi opened the pull request:
Related to issue #849  

In the 'Insert Video from the web' tab, now we get a preview of the video of youtube URLs containing https and youtu.be and the video can be played too using the HTML5 player in the preview tab. :)

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/OpenLiveWriter/OpenLiveWriter?pullRequest=871) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/OpenLiveWriter/OpenLiveWriter?pullRequest=871) before we can accept your contribution.<br/><br/>:x: ranjanjharavi [sign now](https://cla.dotnetfoundation.org/OpenLiveWriter/OpenLiveWriter?pullRequest=871)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/OpenLiveWriter/OpenLiveWriter?pullRequest=871) it.</sub>

nvella commented on src/managed/OpenLiveWriter.PostEditor/Video/VideoProvidersB2.xml:
```diff
@@ -79,28 +79,28 @@
     <serviceName>YouTube</serviceName>
     <editorFormat xpp:if="version lt 15.4.3104.0816">
```
Should there be the duplication of editorFormat and embedFormat entries? They all seem to be describing the same formats.

nvella replied:
A single editorFormat entry and embedFormat entry may do the trick, without the `xpp:if` version specifier. I'm not sure what exactly this does, but I have a hunch it's to do with Adobe Flash versioning.

ranjanjharavi replied:
When I tried to search the URI, XML namespace xpp("http://writer.live.com/xmlpp/2007") it pointed to Windows Essentials. 
It is trying to detect something's version (not sure what) and showing whether to preview normal or HD quality video, based on the version it applies the query params ```?hd=1``` same applies for size element also which is defined just below.

Yes, I tried with single editorFormat and embedFormat without the ```xpp:if``` and it works. Also, in some of the below videoproviders even though size element is used based on the version editor and embedformat is used once. Should I change it to single?

nvella replied:
It does simplify the code a bit by having the single editorFormat and embedFormat entries. Can always be added back later if there seems to be issues 👍

nvella commented on src/managed/OpenLiveWriter.PostEditor/Video/VideoProvidersB2.xml:
```diff
@@ -79,28 +79,28 @@
     <editorFormat xpp:if="version lt 15.4.3104.0816">
       <![CDATA[
-		<embed src="http://www.youtube.com/v/{videoId}" type="application/x-shockwave-flash" wmode="transparent" width="{width}" height="{height}"></embed>
+		<embed src="http://www.youtube.com/embed/{videoId}" wmode="transparent" width="{width}" height="{height}"></embed>
```
We should maybe consider using the same YouTube embedding syntax as provided through the Share button on YouTube.

```
<iframe width="560" height="315" src="https://www.youtube.com/embed/X-a56UVN0cc" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

ranjanjharavi replied:
Yes, Nick, even though there is not much difference in case of video embedding but its better to use <iframe> than <embed>, I also thought to change it to iframe but had doubts about whether to change existing things or not. Better we'll change it to <iframe> :)

nvella commented:
Hi @ranjanjharavi, looks good so far! I've tried out your changes and it seems to work functionality-wise. When I add a video from YouTube though, it does seem to include quite a lot of lag into the editor. I'm not sure what's causing this, it may be unrelated to your changes.

I've left some comments on some things that you may want to try, it may or may not fix the lag. If there's no difference between using `embed` and `iframe` I'll merge these changes anyway, as the problem then would probably be to do with our handling of smart content and not the embedded element specifically.

Thanks again for your contributions! 👍 Nick

ranjanjharavi commented on src/managed/OpenLiveWriter.PostEditor/Video/WebVideoSource.cs:
```diff
@@ -376,7 +376,7 @@ private string CreateEmbedHtml(string embed)
         private static string CreateHtml(string html, string color)
         {
-            return String.Format(CultureInfo.InvariantCulture, "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\" \"http://www.w3.org/TR/html4/strict.dtd\"><body style='margin: 0; padding: 0;background-color: #{1}'>{0}</body></html>", html, color);
+            return String.Format(CultureInfo.InvariantCulture, "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\" \"http://www.w3.org/TR/html4/strict.dtd\"><head><meta content='IE=Edge' http-equiv='X-UA-Compatible'/></head><body style='margin: 0; padding: 0;background-color: #{1}'>{0}</body></html>", html, color);
```
@nvella, I think it would be better if we add HTML5 syntax here instead of older html4. Should I change it to HTML5 syntax?

nvella replied:
Sounds good to me 👍 The HTML5 syntax is much cleaner looking, and should work on all of the platforms we currently support

ranjanjharavi commented:
Hi @nvella, 
Thank you for your valuable feedback. 
I think the lag is because of lots of requests made to youtube API during the embed videos. I agree with you regarding few changes which you pointed out in the comments and I have mentioned my views too. :)

nvella commented:
All good @ranjanjharavi. I've left replies to all the individual comments - pretty much in summary I think it'll be a good idea trying out anything that leaves the code cleaner. Can always go back if stuff breaks.

As for the lag, I'm not quite sure as to exactly what's causing it. Pasting a YouTube iframe embed into a post manually doesn't exhibit the same problem - I think it could be something to do with how Writer is specifically treating the element like a 'video' through the smart content system; this is what displays the Ribbon tab for video tools. I wouldn't worry about the lag too much for now, it'd be something we'd make an issue for later on.

Thanks much for your work so far 😄 Nick

ranjanjharavi commented:
Hi @nvella, Thank you 😄. I have made the above-mentioned changes. Please review it.

nvella commented on src/managed/OpenLiveWriter.PostEditor/Video/VideoProvidersB2.xml:
```diff
@@ -77,30 +77,20 @@
+    <embedFormat>
       <![CDATA[
-      <object width="{width}" height="{height}"><param name="movie" value="http://www.youtube.com/v/{videoId}?hl=en&hd=1"></param><embed src="http://www.youtube.com/v/{videoId}?hl=en&hd=1" type="application/x-shockwave-flash" width="{width}" height="{height}"></embed></object>
+      <object width="{width}" height="{height}"><param name="movie" value="http://www.youtube.com/embed/{videoId}&hl=en"></param><iframe src="http://www.youtube.com/embed/{videoId}&hl=en" width="{width}" height="{height}" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></object>
```
It seems wrapping the `iframe` in the `object` tag breaks videos on Blogger. I think we should be able to remove the tag and just have the `iframe` alone 👍

ranjanjharavi replied:
Sure, I have removed the ```object``` tag, updated the changes and checked it too. 👍

ranjanjharavi replied:
Sorry for not being clear enough I checked that in Wordpress only and there it was working. Also, I was not able to login to blogger while debugging. It was getting 401 error. I think first we have to login through the default installed the application. Or Am I doing it wrong 🤔

nvella commented:
Hi again @ranjanjharavi. I've found one more problem that seems to present when posting YouTube videos to Blogger. Apart from that, I think it's all good for merging 🎉

ranjanjharavi commented:
Hi @nvella. Thanks again for the review and appreciate your valuable feedback. I have removed the ```object``` tag and updated the changes. 😄

nvella commented on src/managed/OpenLiveWriter.PostEditor/Video/VideoProvidersB2.xml:
```diff
@@ -77,30 +77,20 @@
+    <embedFormat>
       <![CDATA[
-      <object width="{width}" height="{height}"><param name="movie" value="http://www.youtube.com/v/{videoId}?hl=en&hd=1"></param><embed src="http://www.youtube.com/v/{videoId}?hl=en&hd=1" type="application/x-shockwave-flash" width="{width}" height="{height}"></embed></object>
+      <iframe src="http://www.youtube.com/embed/{videoId}&hl=en" width="{width}" height="{height}" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```
Seems the `&hl=en` is causing the player to error out (it's not valid URL parameter syntax). It can be changed to `?hl=en`, but I doubt the parameter is needed at all in this case. Removing it so the URL is just `http://www.youtube.com/embed/{videoId}` should be fine.

ranjanjharavi replied:
I tried to follow the same syntax as in the other video providers, and I was trying with WordPress publishing. I agree that we should remove that query parameter. 👍

nvella commented:
I tracked down what was causing the player to break, once it's fixed up it should be good for merge 😃

ranjanjharavi commented:
Hi @nvella, Thanks for figuring it out and I have removed the query parameter and updated the changes.
I think we should clean up for other Video providers too, as they too are using the old ```html``` tags. Thanks again really appreciate your help 😄 🎉

nvella commented:
Looks all good, merged! Thanks for your help 😃

ranjanjharavi commented:
Thank you, Nick 😄
